python/taichi/examples/simulation/alien_life.py

```diff
@@ -0,0 +1,174 @@
+
+SIMPLEX = ti.Vector.field(4, int, shape=64)
+SIMPLEX.from_numpy(simplex_np)
+
+
+@ti.func
+def noise4d(x, y, z, w):
+    """4D simplex noise."""
+    s = (x + y + z + w) * F4
+    ijkl = tm.floor(tm.vec4(x, y, z, w) + s)
+    t = (ijkl.x + ijkl.y + ijkl.z + ijkl.w) * G4
+    xyzw0 = tm.vec4(x, y, z, w) - ijkl + t
+
+    c = int(xyzw0.x > xyzw0.y) * 32 + int(xyzw0.x > xyzw0.z) * 16 + int(
+        xyzw0.y > xyzw0.z) * 8 + int(xyzw0.x > xyzw0.w) * 4 + int(
+            xyzw0.y > xyzw0.w) * 2 + int(xyzw0.z > xyzw0.w)
+    ijkl1 = SIMPLEX[c] >= 3
+    ijkl2 = SIMPLEX[c] >= 2
+    ijkl3 = SIMPLEX[c] >= 1
+
+    xyzw1 = xyzw0 - ijkl1 + G4
+    xyzw2 = xyzw0 - ijkl2 + 2 * G4
+    xyzw3 = xyzw0 - ijkl3 + 3 * G4
+    xyzw4 = xyzw0 - 1 + 4 * G4
+
+    I, J, K, L = ijkl.cast(int) & 255
+    i1, j1, k1, l1 = ijkl1
+    i2, j2, k2, l2 = ijkl2
+    i3, j3, k3, l3 = ijkl3
+    gi0 = PERM[I + PERM[J + PERM[K + PERM[L]]]] & 0x1f
+    gi1 = PERM[I + i1 + PERM[J + j1 + PERM[K + k1 + PERM[L + l1]]]] & 0x1f
+    gi2 = PERM[I + i2 + PERM[J + j2 + PERM[K + k2 + PERM[L + l2]]]] & 0x1f
+    gi3 = PERM[I + i3 + PERM[J + j3 + PERM[K + k3 + PERM[L + l3]]]] & 0x1f
+    gi4 = PERM[I + 1 + PERM[J + 1 + PERM[K + 1 + PERM[L + 1]]]] & 0x1f
+
+    zs = [xyzw0, xyzw1, xyzw2, xyzw3, xyzw4]
+    inds = [gi0, gi1, gi2, gi3, gi4]
+    noise = 0.
+    for i in ti.static(range(5)):
+        t = 0.6 - tm.dot(zs[i], zs[i])
+        if t >= 0:
+            t *= t
+            noise += t * t * tm.dot(zs[i], GRAD4[inds[i]])
+
+    return 27.0 * noise
+
+
+n = 40000
+angs = np.random.uniform(0, 2 * np.pi, n)
+rads = np.sqrt(np.random.uniform(0, 1, n))
+initial_positions = np.zeros((n, 2), dtype=np.float32)
+X, Y = initial_positions[:, 0], initial_positions[:, 1]
+X[...] = rads * np.cos(angs)
+Y[...] = rads * np.sin(angs)
+intensity = np.power(1.001 - np.sqrt(X**2 + Y**2), 0.75)
+pos = ti.Vector.field(2, float, shape=n)
+
+scale = 0.6  # control sync between adjacent points
+length = 0.65  # control amplitude of point wiggles
+
+
+@ti.kernel
+def update(
+    time: float,  # pylint: disable=redefined-outer-name
+    intensity: ti.types.ndarray(),  # pylint: disable=redefined-outer-name
+    points: ti.types.ndarray(dtype=tm.vec2, ndim=1)):
+    pos.fill(0)
+    ct = 1.5 * ti.cos(2 * np.pi * time)
+    st = 1.5 * ti.sin(2 * np.pi * time)
+    for i in range(n):
+        x, y = points[i]
+        dx = noise4d(scale * x, scale * y, ct, st)
+        dx *= intensity[i] * length
+        dy = noise4d(100 + scale * x, 200 + scale * y, ct, st)
+        dy *= intensity[i] * length
+        pos[i] = x + dx, y + dy
+
+
+size = 800
+gui = ti.GUI("Alien Life", res=(size, size))
+time = 0
+theta = np.linspace(0, 2 * np.pi, 200)
+begin = np.array([np.cos(theta)[:-1], np.sin(theta[:-1])]).transpose()
+end = np.array([np.cos(theta)[1:], np.sin(theta[1:])]).transpose()
+begin = (begin + 1) / 2
+end = (end + 1) / 2
+
+while gui.running:
+    time += 1
+    update(time * 0.002, intensity, initial_positions)
```

Comment:
This makes a copy of `intensity` and `initial_positions` every time we call the kernel. Although they are GCed, we shouldn't encourage this usage because it hinders performance